### ISSUE TITLE
fbmuck creates log files with incorrect ownership

### ISSUE THREAD

**Issue by hyperlumic** [fuzzball-muck/fuzzball#236]
fbmuck (compiled from master at 2018-11-30 09:30 PM EST) with the following specified in `include/config.h`:

```
#define MUD_ID "fuzzball"
#define MUD_GID "fuzzball"
```

fbmuck launched from start script as root user creates logfiles that runtime UID/GID cannot access, and silently fails. It looks like the logfiles are created before privileges are dropped and the process UID is switched. As a result, the new effective UID can't read the logfiles, and bails.

```
root@sprdev01:~ # /home/fuzzball/spr/restart

root@sprdev01:~ # ps -aux | grep fbmuck
root     41896   0.0  0.0   408   324  0  R+   21:39    0:00.00 grep fbmuck

root@sprdev01:~ # ls -l /home/fuzzball/spr/game/logs/
total 12
-rw-------  1 root  fuzzball    0 Nov 30 21:39 fbmuck
-rw-------  1 root  fuzzball   28 Nov 30 21:39 fbmuck.err
-rw-------  1 root  fuzzball   79 Nov 30 21:39 restarts
-rw-------  1 root  fuzzball  119 Nov 30 21:39 status

root@sprdev01:~ # cat /home/fuzzball/spr/game/logs/fbmuck.err
Unable to open logs/status!
```

Platform: `FreeBSD sprdev01.sprmuck.org 11.2-RELEASE-p5 FreeBSD 11.2-RELEASE-p5 #0: Tue Nov 27 09:33:52 UTC 2018     root@amd64-builder.daemonology.net:/usr/obj/usr/src/sys/GENERIC  amd64`

**Comment by tanabi** [comment]
@wyld-sw @aidanPB @starafar

Hey guys -- I wanted some opinions on how to fix this.

Here's the problem -- in Fuzzball's main(...) function one of the first things it does is open file handles for its messages and redirect stdout/stderr to go to those files.  If you are running the MUCK as root with the intent to setuid to a user after the MUCK is booted (i.e. to open the muck on a protected port) this will create the files as root.

I've come up with two solutions, both have potential issues.  I'm not sure which issues would be considered worse which is why I'm soliciting some opinions.  If anyone has a better idea, let me know, but I'm pretty sure there's only 2 ways to do this.

Way 1: Open the log files AFTER we open the ports.  I would probably shift main(...) around a little to open the ports a little earlier (before the DB loads) because any errors that are produced won't be logged but instead will go to stderr.

Downside is, if there's a problem with the ports -- for instance, port already in use by another program -- the error message goes to stderr rather than the file.  If you're using something like systemd or a Docker container for the MUCK, this isn't a huge deal since the message is still accessible albiet in a different place from all the other fuzzball errors.  If you're using something like rc.local to kick off the MUCK, though, the message goes off into the ether.

Way 2: Change file ownership on the files to match MUCK_ID/MUCK_GID after opening them so they don't get stuck as root.  Perhaps we check the ownership first and only change them if the ownership is root, so that we're not changing around file permissions for no reason.

Downside is, you lose the ability to control the ownership of those files.  If there's some reason you want that ownership to be set to root, it's not going to stick.  My programmer sense tells me this is 'hacky' though so I'm not in love with it, but it seems a little more 'seamless'?  There's a definite seam in Way 1 since there are potential messages not being sent to the same place as literally every other message.

Any thoughts / preferences, guys?

**Comment by wyld-sw** [comment]
Agreed on your assessment of the options. For option 1, would it solve anything to "hold" any port error messages until logs are open, then process them? Or would that be even more hacky?

Definitely not in my wheelhouse these past couple years, so I might be missing something crucial. I certainly am interested in other opinions! @aindanPB @starafar

**Comment by tanabi** [comment]
@wyld-sw Hrm, that's a 'permutation' I hadn't thought about.  I could defer those log messages, or make a log-opener function that will do the setuid stuff before trying to open the files and make the port opening code open the log file if it needs to.

I'm thinking the latter idea (log opener function, open log file if needed in connection code) is perhaps the 'right' way to do it.

**Comment by wyld-sw** [comment]
@tanabi Log opener sounds great.

**Comment by PointsDragon** [comment]
Issue here is if you drop permissions and THEN try and open your log files - and now have no permissions to create the files at all.  Now you're in a really bad space for output.  My take is that if someone sets ID and/or GID, the server should:

1. Do the file(s) needed already exist?  Can I read/write to them with the requested UID/GID?  If so, use them.  STOP.
2. Do files I can't read/write exist with names I need?  Attempt to move them.  If I'm unable to move files, ERROR.
3. Create new files, set expected UID/GID.  STOP.

Then we document in config.h that THIS is the behavior you are going to get.

**Comment by wyld-sw** [comment]
I'm good with that approach.

**Comment by tanabi** [comment]
@PointsDragon Seems sensible to me as well.

**Comment by PointsDragon** [comment]
Note - we have one file we always open, and up to two files we potentially open.  PID file is always opened.  We only re-direct via freopen stderr/stdout if we're not in interactive mode.  Additionally, this can get a bit more complex in that the original shell may have redirected one or both streams together, and the admin may have set the redirect filenames to the same file.  Not sure how much we need to track these various paths down.  However, for the decision tree noted above, we'll need to track 'up to' three targets for state.
